yt_dlp/extractor/viu.py

```diff
@@ -271,27 +286,41 @@ def _real_extract(self, url):
         url, idata = unsmuggle_url(url, {})
         country_code, lang_code, video_id = self._match_valid_url(url).groups()
 
+        webpage = self._download_webpage(url, video_id, fatal=False)
+        json_ld = self._search_json_ld(webpage, video_id, fatal=False)
+        next_js_data = (self._search_nextjs_data(webpage, video_id, fatal=False) or {}).get('props')
+        runtime_info = traverse_obj(next_js_data, ('initialState', 'app', 'runtimeInfo'))
+
         query = {
             'r': 'vod/ajax-detail',
             'platform_flag_label': 'web',
             'product_id': video_id,
         }
 
-        area_id = self._AREA_ID.get(country_code.upper())
+        area_id = self._AREA_ID.get(country_code.upper()) or runtime_info.get('areaId')
         if area_id:
             query['area_id'] = area_id
 
-        product_data = self._download_json(
-            f'http://www.viu.com/ott/{country_code}/index.php', video_id,
-            'Downloading video info', query=query)['data']
+        try:
+            product_data = self._download_json(
+                f'http://www.viu.com/ott/{country_code}/index.php', video_id,
+                'Downloading video info', query=query, fatal=False)['data']
+        # The `fatal` in `_download_json` didn't prevent json error
+        # FIXME: probably the error still too broad
+        except (DownloadError, ExtractorError):
+            # NOTE: some geo-blocked like https://www.viu.com/ott/sg/en/vod/108599/The-Beauty-Inside actually can bypassed
+            # on other region (like in ID)
+            product_data = traverse_obj(
+                next_js_data, ('pageProps', 'fallback', lambda k, v: v if re.match(r'@"PRODUCT_DETAIL"[^:]+', k) else None),
+                get_all=False)['data']
```

Comment:
Any suggestion about this? I think `DownloadError` or `ExtractorError` scope is too broad and may set wrong handling of the error

Reply:
`ExtractorError` only is sufficient.

Am I understanding correctly that you only want to catch a JSON error here? If so, then you can do something like this:
```py
        except ExtractorError as e:
            if not isinstance(e.cause, json.JSONDecodeError):
                raise
            # fallback code
```

Reply:
Hmm, still not handling error correctly. Here's the new code:
```python
try:
    product_data = self._download_json(
        f"http://www.viu.com/ott/{country_code}/index.php",
        video_id,
        "Downloading video info",
        query=query,
        fatal=False,
        expected_status=404,
    )["data"]
# The `fatal` in `_download_json` didn't prevent json error
# FIXME: probably the error still too broad
except ExtractorError as e:
    if not isinstance(e.cause, json.JSONDecodeError):
        raise
    # NOTE: some geo-blocked like https://www.viu.com/ott/sg/en/vod/108599/The-Beauty-Inside actually can bypassed
    # on other region (like in ID)
    product_data = traverse_obj(
        next_js_data,
        (
            "pageProps",
            "fallback",
            lambda k, v: v if re.match(r'@"PRODUCT_DETAIL"[^:]+', k) else None,
        ),
        get_all=False,
    )["data"]

```
This code give error like this:
```
yt_dlp.utils.DownloadError: ERROR: [viu:ott] 2221644: [viu:ott] 2221644: Failed to parse JSON: Expecting value in '': line 1 column 1 (char 0); please report this issue on  https://github.com/yt-dlp/yt-dlp/issues?q= , filling out the appropriate issue template. Confirm you are on the latest version using  yt-dlp -U
```

I want to get `product_json` json if available (this data available in other region). If not, i want to parse from webpage. `_download_json` still raise error when `fatal=False` with `expected_status=404` (this endpoint not available in `ID`). I waill try use `DownloadError`

Reply:
Are you sure you're not just getting a warning (since `fatal=False`) and then the download test is raising as an error?

In order to actually catch the error, the `_download_json` call needs to be fatal (and not expecting 404). Try removing `fatal=False` and `expected_status=404`

Reply:
The error handling fixed in f92dedc92a3356958f986ffc145e61a39d554945